Firestore/core/src/firebase/firestore/local/query_engine.h

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_QUERY_ENGINE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_QUERY_ENGINE_H_
+
+#include "Firestore/core/src/firebase/firestore/core/query.h"
```

Comment:
We've seen nontrivial compilation speedups by using forward declarations in headers. To the extent that this is possible, we should do so. Not all existing code has been migrated to this style yet.

In general the compiler needs to know the full declaration of any non-pointer members of classes or any base classes. Otherwise concrete types can be forward declared. Note that type aliases generally can't be forward declared and have to be included.

In this case you can forward declare LocalDocumentsView, Query, and SnapshotVersion. DocumentKeySet can't be forward declared because it's a type alias.

See [`sizer.h`](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/core/src/firebase/firestore/local/sizer.h) for an example of what this looks like with differing namespaces.

Reply:
I thought it was bad style to use forward declarations and that it is better to include the relevant headers. The build time argument makes sense to me and seems like a worthwhile tradeoff.

I added the forward declarations to query_engine.h, but not to simple_query_enigine.h. Is it better to have all dependencies spelled out in each individual header file?